### PR TITLE
Distinguish between address and contents uniformity

### PR DESCRIPTION
**Pull request by alan-baker**
Fixes https://github.com/gpuweb/gpuweb/issues/3677

* Add new parameter tag for content uniformity
* Update analysis to distinguish between contents and addresses

Just look at the commits starting at https://github.com/gpuweb/gpuweb/commit/ca6f515d13e10fa67aa19fbe7b1eaefe456d50d5.

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -9879,10 +9879,18 @@ For each [=formal parameter=] of a function, one or two tags are computed:
   <thead>
     <tr><th>Parameter Tag<th>Description
   </thead>
-  <tr algorithm="ParameterRequiredToBeUniform tag">
+  <tr algorithm="ParameterRequiredToBeUniformValue tag">
       <td><dfn noexport>ParameterRequiredToBeUniform.S</dfn>,<br>
           where |S| is one of the severities: [=severity/error=], [=severity/warning=], or [=severity/info=].
       <td>The parameter must be a [=uniform value=].
+          If the parameter type is a pointer, the memory view must be uniform.
```
Comment (dneto0) [inline wgsl/index.bs:9886]: Can we indicate somehow that we're not making a judgement about the contents of the referenced locations?

> If the parameter is a pointer, the memory view, but not necessarily its contents, must be uniform.

File: wgsl/index.bs [hunk 2]
```diff
@@ -9879,10 +9879,18 @@ For each [=formal parameter=] of a function, one or two tags are computed:
...
+          Otherwise a diagnostic with severity |S| will be triggered.
+
+          Associated with a [=witness set=].
+  <tr algorithm="ParameterContentsRequiredToBeUniform tag">
+      <td><dfn noexport>ParameterContentsRequiredToBeUniform.S</dfn>,<br>
+          where |S| is one of the severities: [=severity/error=], [=severity/warning=], or [=severity/info=].
+      <td>The contents of the pointer parameter must be a [=uniform value=].
```
Comment (dneto0) [inline wgsl/index.bs:9893]: Suggest: 

> The parameter is a pointer, and the contents of its memory view must be a [=uniform value=].

Or alternately mirror the description of PointerParameterMayBeNonUniform 

> The value stored in the memory pointed to by the pointer parameter must be a [=uniform value=].

**Review by dneto0 — APPROVED** [review]
LGTM with nits about clarity of wording.

Thanks!

**Review by jrprice — COMMENTED** [review]
I see that you didn't make the distinction for `ParameterReturnRequiredToBeUniform`. That might be OK for V1 due to the various restrictions we have on pointers, but I think we should at least add clarification to that tag description to highlight that the requirement for a pointer is on the contents. Or just split the tag like you did for the direct parameter requirement, which will be necessary when we have better pointers in the future anyway.

**Comment by dneto0** [comment]
Waiting for @jrprice review

File: wgsl/index.bs [hunk 3]
```diff
@@ -9805,8 +9814,8 @@ For each [=formal parameter=] of a function, one or two tags are computed:
   <thead>
     <tr><th>Parameter Return Tag<th>Description
   </thead>
-  <tr><td><dfn noexport>ParameterReturnRequiredToBeUniform</dfn>
-      <td>The parameter [=shader-creation error|must=] be a [=uniform value=] in order for the [=return value=] to be a uniform value.
+  <tr><td><dfn noexport>ParameterReturnContentsRequiredToBeUniform</dfn>
+      <td>The parameter [=shader-creation error|must=] be a [=uniform value=] in order for the [=return value=] contents to be a uniform value.
```
Comment (jrprice) [inline wgsl/index.bs:9818]: I think the "contents" is in the wrong place (should relate to the parameter, not the return value). This should actually be "the parameter contents must be a uniform value in order for the return value to be a uniform value".

Comment (dneto0) [reply wgsl/index.bs:9818]: Would you change the description too?

> The parameter must be a uniform value in order for the return value to be a uniform value.
If the parameter is a pointer, then additionally value stored in the memory pointed to by the parameter must be uniform in order for the return value to be uniform.

Comment (jrprice) [reply wgsl/index.bs:9818]: Yes, my comment was more about the description than the tag name itself.

FWIW Tint uses two separate tags for pointer value vs pointer contents here, which is not necessary for V1 but will be required when we remove restrictions on pointer parameters.

**Review by dneto0 — APPROVED** [review]
LGTM, but @jrprice must review

File: wgsl/index.bs [hunk 4]
```diff
@@ -9811,8 +9820,9 @@ For each [=formal parameter=] of a function, one or two tags are computed:
...
```
Comment (jrprice) [inline wgsl/index.bs:9824]: Nit: I think we should drop the "contents" here from "return value contents", since the return value is not a pointer.

**Review by jrprice — APPROVED** [review]

**Pull request merged**